Техническое задание №7

AndreyMMGH opened the pull request.

AndreyMMGH committed:
Добавить перечисление TaskType с типами задач

AndreyMMGH committed:
Добавить вспомогательный утилитарный класс CSVTaskFormat и метод getTaskType в задачах

AndreyMMGH committed:
Добавить собственное непроверяемое исключение ManagerSaveException

AndreyMMGH committed:
Создать класс FileBackedTaskManager с возможностью сохранения, восстановления задач, эпиков, подзадач и истории

AndreyMMGH committed:
Скорректировать методы

AndreyMMGH committed:
Изменить идентификаторы доступа у HashMap, generatorId и historyManager

AndreyMMGH committed:
Сделать корректировки в методе main и getDefault для загрузки и выгрузки данных в файл .csv

AndreyMMGH committed:
Сделать изменение в классе теста InMemoryTasksManagerTest. Заменить на new InMemoryTasksManager()

AndreyMMGH committed:
Добавить новы класс по тестам FileBackedTaskManagerTest с 2-мя тестами:
- сохранение и загрузку пустого файла;
- сохранение нескольких задач;
- загрузку нескольких задач.

AndreyMMGH committed:
Переделать import в Main

AndreyMMGH committed:
Сделать выравнивание кода в классе Task

AndreyMMGH committed:
Переименовать переменную в соответствии LowerCamelCase

kesch9 commented on src/service/CSVTaskFormat.java:
```diff
@@ -0,0 +1,98 @@
+
+public class CSVTaskFormat {
+
+    public static String toString(Task task) {
```
Этот метод можно упростить, для Task и Epic запись одинаковая,  можно объединить. Последний else лишний, т.к. у вас метод не может не вернуть тип задач. Поэтому можно сделать одну проверку на Subtask и если не подзадача возвращаем без  epicId, иначе с ним

AndreyMMGH replied:
Оптимизировал метод. Убрал лишний код

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+                    Subtask subtask = managerForRecovery.getSubtask(taskId);
+
+                    if (task != null || epic != null || subtask != null) {
+                        managerForRecovery.historyManager.addTask(task);
```
Здесь ошибка, т.к. всегда добавляется task, а нужно в зависимости, что не null, тот тип задачи и добавлять. Например если epic != null, то его добавляем

AndreyMMGH replied:
Сделал несколько условий для задачи, эпика и подзадачи

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+                List<Integer> idsHistory = CSVTaskFormat.fromStringToHistoryId(idsHistoryLine);
+
+                for (Integer taskId : idsHistory) {
+                    Task task = managerForRecovery.getTask(taskId);
```
Нельзя вызывать метод getTask т.к. он должен предопределятся и вызываться метод save, нужно напрямую получить задачу из хранилища, для остальных типов также

AndreyMMGH replied:
Переделал получение задач разных типов напрямую из хранилища

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+        }
+
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(file, StandardCharsets.UTF_8))) {
+            writer.write("id,type,name,status,description,epic");
```
Можно вынести header в константы

AndreyMMGH replied:
Вынес шапку в константу. Вроде, константу нужно большими буквами писать. Тест не пропускал на гитхабе

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+            writer.newLine();
+
+
+            writer.write(CSVTaskFormat.fromHistoryIdToString(historyManager));
```
Писать историю и восстанавливать не нужно было, но если сделали надо подправить замечания по нему

AndreyMMGH replied:
В рекомендательном видео у наставника описывалась ситуация с историей. Решил попробовать сделать. Получилось криво)

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+    public void deleteSubtasks() {
+        super.deleteSubtasks();
+        save();
+    }
```
Т.к. вы сохраняете историю, нужно еще переопределить метод получения задач по id, где происходит запись в историю, вам нужно сохранять актуальную историю при добавлении задач в историю

AndreyMMGH replied:
Переопределил 3 метода getTask(int id), getEpic(int id) и getSubtask(int id)

kesch9 commented on test/service/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,76 @@
+
+        FileBackedTaskManager subsequentManager = FileBackedTaskManager.loadFromFile(file);
+
+        assertEquals(initialManager.getTask(idSnowRemoval), subsequentManager.getTask(idSnowRemoval), "Задача snowRemoval из разных менеджеров должна совпасть");
```
Т.к. equals и hashcode считается только по id у вас здесь проверяется равенство по id, вам же нужно проверить, что все поля восстановились верно, поэтому нужно каждое поле проверить. Для подзадач и эпиков нужно учесть их специфичные поля

AndreyMMGH replied:
Постарался сделать проверку по полям. Во время написания теста обнаружил у себя неточности в конструкторе эпика и подзадачи

kesch9 reviewed (commented):
Нужно еще поработать над заданием

AndreyMMGH committed:
Оптимизировать метод toString в классе CSVTaskFormat

AndreyMMGH committed:
Оптимизировать №2 метод toString в классе CSVTaskFormat

AndreyMMGH committed:
Исправить ошибки:
- добавление эпика и подзадачи;
- получение задачи, эпика и подзадачи из хранилища;
- переопределить методы получения задач по id.

AndreyMMGH committed:
Сделать константу для шапки

AndreyMMGH committed:
Поменять местами name и description в конструкторе

AndreyMMGH committed:
В description эпика сделать запись без запятой

AndreyMMGH committed:
Переделать тест saveAndRestoreTasks, проверяя все поля у задач, эпиков и подзадач

AndreyMMGH committed:
Изменить написание константы

AndreyMMGH committed:
Изменить написание константы. Вариант №2

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,221 @@
+public class FileBackedTaskManager extends InMemoryTasksManager {
+
+    private final File file;
+    final String header = "id,type,name,status,description,epic";
```
Для констант нужно указывать static и UPPER_CASE пишутся, также не хватает private
private final static String HEADER

kesch9 reviewed (approved):
Approve
